<a href="https://colab.research.google.com/github/LeeSeungwon89/Machine-learning_Theory/blob/master/CHAPTER8%20%ED%85%8D%EC%8A%A4%ED%8A%B8%20%EB%B6%84%EC%84%9D.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **NLP이냐 텍스트 분석이냐**

기실 NLP와 텍스트 분석을 구분하는 것은 큰 의미가 없지만 굳이 나눈다면 아래와 같습니다.

**NLP(National Language Processing, 자연어 처리)**는 머신이 인간의 언어를 이해하고 해석하는 데 더 중점을 두고 발전하고 있습니다. NLP의 영역은 언어를 해석하기 위한 기계 번역, 자동으로 질문을 해석하고 답을 해주는 질의응답 시스템 등입니다.

**TA(Text Analytics, Text Mining, 텍스트 분석)**은 비정형 텍스트에서 의미 있는 정보를 추출하는 데 중점을 두고 발전하고 있습니다. NLP는 TA의 기반 기술이며, NLP가 발전하면서 TA도 더 정교하게 발전했습니다. 

머신러닝 기술은 NLP와 TA의 발전에 크게 기여했습니다. 예전의 텍스트를 구성하는 언어적인 룰이나 업무의 룰에 따라 텍스트를 분석하는 룰 기반 시스템에서 머신러닝의 텍스트 데이터를 기반으로 모델을 학습하고 예측하는 기반으로 변경되면서 기술이 비약적으로 발전한 것입니다.

TA는 머신러닝, 언어 이해, 통계 등을 확용하여 모델을 수립하고 정보를 추출하여 비즈니스 인텔리전스나 예측 분석 등 분석 작업을 주로 수행합니다. 주로 아래와 같은 기술 영역에 집중합니다.

- **텍스트 분류(Text Classification, Text Categorization)**: 문서가 특정 분류 또는 카테고리에 속하는 것을 예측하는 기법을 통칭합니다. 지도학습을 적용하며, 신문 기사가 특정 카테고리에 속하는지 자동으로 분류하거나 스팸 메일을 검출하는 기능을 예로 들 수 있습니다. 

- **감성 분석(Sentiment Analysis)**: 텍스트에서 나타나는 주관적인 요소를 분석하는 기법을 총칭합니다. 지도학습과 비지도학습을 모두 적용할 수 있으며, 소셜 미디어 감정 분석, 영화나 제품에 대한 리뷰, 여론조사 의견 분석 등을 예로 들 수 있습니다. 

- **텍스트 요약(Summarization)**: 텍스트 내에서 중요한 주제나 중심 사상을 추출하는 기법을 의미합니다. 토픽 모델링(Topic Modeling)을 예로 들 수 있습니다.

- **텍스트 군집화**와 **유사도 측정**: 비슷한 유형의 문서를 군집화하는 기법을 의미합니다. 텍스트 분류를 비지도학습으로 수행하는 방법의 일환으로 사용될 수 있습니다. 유사도 측정 역시 문서들 간 유사도를 측정하여 비슷한 문서끼리 모을 수 있는 기법입니다.

# **1. 텍스트 분석 이해**

지금까지 ML 모델은 정형 데이터 기반에서 모델을 수립하고 예측을 수행했습니다. 그리고 머신러닝 알고리즘은 숫자형 피처 기반 데이터만 입력받을 수 있으므로 텍스트를 머신러닝에 적용하려면 비정형 텍스트 데이터를 피처 형태로 추출하고 추출된 피처에 의미 있는 값을 부여하는 것이 중요합니다. 텍스트를 단어 기반(또는 단어 일부분)의 다수 피처로 추출하고 이 피처에 단어 빈도수 같은 숫자 값을 부여하면 텍스트는 단어 조합인 벡터값으로 표현될 수 있습니다. 이렇게 텍스트를 변환하는 것을 피처 벡터화(Feature Vectorization) 또는 피처 추출(Feature Extraction)이라고 합니다. 텍스트를 피처 벡터화 하여 변환하는 방법은 **BOW(Bag of Words)**와 **Word2Vec**가 대표적입니다.


## **1.1. 텍스트 분석 수행 프로세스**

TA는 아래와 같은 프로세스 순으로 수행합니다.

1. 텍스트 사전 준비(텍스트 전처리): 클렌징, 대/소문자 변경, 특수문자 삭제, 단어 등의 토큰화, 무의미한 단어(Stop word) 제거, 어근 추출(Stemming/Lemmatization) 등 텍스트 정규화 작업을 수행합니다.

1. 피처 벡터화 및 추출: 전처리로 가공된 텍스트에서 피처를 추출하고 벡터 값을 할당합니다. BOW와 Word2Vec가 대표적인 방법입니다. BOW는 Count 기반과 TF-IDF 기반 벡터화를 수행합니다. 

1. ML 모델 수립 및 학습, 예측, 평가: 피처 벡터화가 수행된 데이터 세트에 ML 모델을 적용하여 학습, 예측, 평가를 수행합니다.


## **1.2. 파이썬 기반의 NLP, 텍스트 분석 패키지**

대표적으로 사용되는 패키지는 아래와 같습니다.

- **NLTK(Natural Language Toolkit for Python)**: 파이썬의 가장 대표적인 NLP 패키지입니다. 방대한 데이터 세트와 서브 모듈을 가졌으며 NLP의 대부분 영역을 커버합니다. 수많은 NLP 패키지가 NLTK의 영향을 받아 작성되고 있습니다. 느린 수행 속도로 인해 대량 데이터 기반에서는 활용되지 않습니다. 아울러 수행 성능, 정확도, 신기술, 엔터프라이즈 기능 지원 등 전반적인 측면이 부족하기 때문에 실무에서 잘 활용되지 않습니다.

- **Genism**: 토픽 모델링 분야에서 가장 두각을 보이며 많이 사용되는 패키지입니다. 토픽 모델링을 쉽게 구현하는 기능을 지원하며, Word2Vec 구현 등 다양한 신기능도 지원합니다.

- **SpaCy**: 수행 성능이 매우 뛰어나 최근들어 주목 받는 패키지입니다. NLP 애플리케이션에서 사용하는 사례가 느는 추세입니다.

참고로 사이킷런은 머신러닝 위주의 라이브러리입니다. 따라서 NLP를 위한 라이브러리, 예컨대 '어근 처리'를 수행하는 것처럼 NLP 패키지에 특화된 라이브러리는 없습니다. 다만 텍스트를 일정 수준으로 가공하고 머신러닝 알고리즘에 텍스트 데이터를 피처로 처리하기 위한 편리한 기능을 제공하므로 충분히 TA를 수행할 수 있습니다. 더 다양한 TA가 적용되어야 할 경우에는 위에서 설명한 패키지를 결합하여 애플리케이션을 작성합니다.

# **2. 텍스트 사전 준비 작업(텍스트 전처리) - 텍스트 정규화**

텍스트를 바로 피처화할 수 없으므로 데이터 전처리 과정을 거쳐야 합니다. 텍스트 정규화는 텍스트를 머신러닝 알고리즘이나 NLP 애플리케이션에 입력 데이터로 사용할 목적으로 클렌징, 정제, 토큰화, 어근화 등 사전 작업을 수행하는 작업입니다. 


## **2.1. 클렌징(Cleansing)**

TA를 수행하면서 불필요하거나 방해 요소로 작용할 만한 문자와 기호(예컨대 HTML, XML 태그나 특정 기호) 등을 사전에 제거하는 작업입니다. 

## **2.2. 텍스트 토큰화(Tokenization)**

텍스트 토큰화 유형은 문서에서 문장을 분리하는 문장 토큰화, 문장에서 단어를 토큰으로 분리하는 단어 토큰화로 나뉩니다. 

### **2.2.1. 문장 토큰화** 

문장 토큰화는 문장의 마침표나 개행문자 등 문장의 마지막을 뜻하는 기호에 따라 분리하는 것이 일반적입니다. 아울러 정규 표현식에 따른 문장 토큰화도 가능합니다. NTLK에서 일반적으로 많이 쓰이는 `sent_tokenize` 클래스로 토큰화해 보겠습니다. 샘플 텍스트 문서를 문장 3개로 분리하는 작업입니다.

In [ ]:
from nltk import sent_tokenize
import nltk
# 마침표, 개행 문자 등의 데이터 세트를 다운로드합니다.
nltk.download('punkt')

text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
You can see it out your window or on your television. \
You feel it when you go to work, or go to church or pay your taxes.'

sentences = sent_tokenize(text=text_sample)
print(type(sentences), len(sentences))
print(sentences)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
<class 'list'> 3
['The Matrix is everywhere its all around us, here even in this room.', 'You can see it out your window or on your television.', 'You feel it when you go to work, or go to church or pay your taxes.']


리스트 객체에 문장 3개가 담겨 있습니다.

### **2.2.2. 단어 토큰화**

단어 토큰화는 문장을 단어로 토큰화하는 작업입니다. 공백, 콤마, 마침표, 개행문자 등으로 단어를 분리하며, 정규 표현식을 이용하여 다양한 유형으로도 토큰화할 수 있습니다. 마침표나 개행문자처럼 문장을 분리하는 구분자를 이용하여 단어를 토큰화할 수 있으므로 Bag of Word처럼 단어 순서가 중요하지 않으면 문장 토큰화를 사용하지 않고 단어 토큰화만 사용해도 충분합니다. 일반적으로 문장 토큰화는 각 문장이 가지는 의미가 중요한 요소로 사용될 때 사용합니다. NLTK의 `word_tokenize()` 클래스를 활용해 보겠습니다.

In [ ]:
from nltk import word_tokenize

sentence = 'The Matrix is everywhere its all around us, here even in this room.'
words = word_tokenize(sentence)
print(type(words), len(words))
print(words)

<class 'list'> 15
['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.']


문장 토큰화와 단어 토큰화를 동시에 수행해 보겠습니다. 먼저 여러 문장을 문장별로 단어 토큰화 하는 함수를 선언하겠습니다.

In [ ]:
def tokenize_text(text):
    sentences = sent_tokenize(text)
    word_tokens = [word_tokenize(sentence) for sentence in sentences]
    return word_tokens

이 함수를 사용하여 두 작업을 한 번에 수행해 보겠습니다.

In [ ]:
text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
You can see it out your window or on your television. \
You feel it when you go to work, or go to church or pay your taxes.'

word_tokens = tokenize_text(text_sample)
print(type(word_tokens), len(word_tokens))
print(word_tokens)

<class 'list'> 3
[['The', 'Matrix', 'is', 'everywhere', 'its', 'all', 'around', 'us', ',', 'here', 'even', 'in', 'this', 'room', '.'], ['You', 'can', 'see', 'it', 'out', 'your', 'window', 'or', 'on', 'your', 'television', '.'], ['You', 'feel', 'it', 'when', 'you', 'go', 'to', 'work', ',', 'or', 'go', 'to', 'church', 'or', 'pay', 'your', 'taxes', '.']]


문장을 단어별로 토큰화하면 문맥 의미를 잃게 됩니다. 이 문제를 다소 해결하는 방안으로 [n-gram](https://wikidocs.net/21692)을 사용합니다. n-gram은 연속된 n개의 단어를 하나의 토큰화 단위로 분리합니다. n개 단어 크기 윈도우를 만들어 문장 처음부터 오른쪽으로 움직이면서 토큰화합니다. 예컨대 'Agent Smith knocks the door'를 2-gram(bigram)으로 만들면 (Agent, Smith), (knocks, the), (the, door)와 같이 연속적으로 단어 2개들을 차례로 이동하면서 토큰화합니다.

## **2.3. 스톱 워드(Stop Word) 제거**

스톱 워드는 문맥적으로 무의미하여 분석에 불필요한 단어를 말합니다. 'is', 'the', 'a', 'will' 등을 예로 들 수 있습니다. 이런 단어들을 제거해야 데이터 분석을 수행할 때 중요 단어로 인지되지 않습니다.

스톱 워드는 언어별로 목록화되어 있습니다. 아래에서 NLTK의 스톱 워드 종류를 확인해 보겠습니다.

In [ ]:
import nltk
nltk.download('stopwords')

print('영어의 스톱 워드 개수: {}개'.format(len(nltk.corpus.stopwords.words('english'))))
print(nltk.corpus.stopwords.words('english')[:10])

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
영어의 스톱 워드 개수: 179개
['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're"]


유의미한 단어만 추출해 보겠습니다.

In [ ]:
text_sample = 'The Matrix is everywhere its all around us, here even in this room. \
You can see it out your window or on your television. \
You feel it when you go to work, or go to church or pay your taxes.'

word_tokens = tokenize_text(text_sample)

stopwords = nltk.corpus.stopwords.words('english')
all_tokens = []

for sentence in word_tokens:
    filtered_words = []
    for word in sentence:
        # 단어를 소문자로 변환합니다.
        word = word.lower()
        # 단어가 스톱 워드 리스트에 없으면 `filtered_words` 리스트에 추가합니다.
        if word not in stopwords:
            filtered_words.append(word)
    # 필터링한 단어를 모은 `filtered_words` 리스트를 `all_tokens` 리스트에 추가합니다.
    all_tokens.append(filtered_words)

print(all_tokens)

[['matrix', 'everywhere', 'around', 'us', ',', 'even', 'room', '.'], ['see', 'window', 'television', '.'], ['feel', 'go', 'work', ',', 'go', 'church', 'pay', 'taxes', '.']]


## **2.4. Stemming과 Lemmatization**

**Stemming**과 **Lemmatization**은 문법적 또는 의미적으로 변화하는 단어의 원형을 찾습니니다. Stemming은 원형 단어로 변환할 때 일반적이고 더 단순한 방법을 적용하여 원래 단어에서 일부 철자가 훼손된 어근 단어를 추출합니다. 반면 Lemmatization은 문법적인 요소와 더 의미적인 부분을 감안하여 정확한 철자로 된 어근 단어를 찾습니다. 따라서 Lemmatization이 더 정교하며 의미론적인 기반에서 단어의 원형을 찾습니다. 다만 수행 시간이 오래 걸립니다.

NLTK가 가진 Stemming을 위한 대표적 Stemmer는 `stem` 패키지의 `porter`, `lancaster`, `snowball` 모듈입니다.

In [ ]:
import nltk.stem

help(nltk.stem)

Help on package nltk.stem in nltk:

NAME
    nltk.stem - NLTK Stemmers

DESCRIPTION
    Interfaces used to remove morphological affixes from words, leaving
    only the word stem.  Stemming algorithms aim to remove those affixes
    required for eg. grammatical role, tense, derivational morphology
    leaving only the stem of the word.  This is a difficult problem due to
    irregular words (eg. common verbs in English), complicated
    morphological rules, and part-of-speech and sense ambiguities
    (eg. ``ceil-`` is not the stem of ``ceiling``).
    
    StemmerI defines a standard interface for stemmers.

PACKAGE CONTENTS
    api
    arlstem
    isri
    lancaster
    porter
    regexp
    rslp
    snowball
    util
    wordnet

FILE
    /usr/local/lib/python3.7/dist-packages/nltk/stem/__init__.py




아래에서 사용할 `LancasterStemmer` 클래스는 `lancaster` 모듈에 포함됩니다.

In [ ]:
help(nltk.stem.lancaster)

Help on module nltk.stem.lancaster in nltk.stem:

NAME
    nltk.stem.lancaster

DESCRIPTION
    A word stemmer based on the Lancaster (Paice/Husk) stemming algorithm.
    Paice, Chris D. "Another Stemmer." ACM SIGIR Forum 24.3 (1990): 56-61.

CLASSES
    nltk.stem.api.StemmerI(builtins.object)
        LancasterStemmer
    
    class LancasterStemmer(nltk.stem.api.StemmerI)
     |  LancasterStemmer(rule_tuple=None, strip_prefix_flag=False)
     |  
     |  Lancaster Stemmer
     |  
     |      >>> from nltk.stem.lancaster import LancasterStemmer
     |      >>> st = LancasterStemmer()
     |      >>> st.stem('maximum')     # Remove "-um" when word is intact
     |      'maxim'
     |      >>> st.stem('presumably')  # Don't remove "-um" when word is not intact
     |      'presum'
     |      >>> st.stem('multiply')    # No action taken if word ends with "-ply"
     |      'multiply'
     |      >>> st.stem('provision')   # Replace "-sion" with "-j" to trigger "j" set of rules
     |   

Lemmatization을 위해서는 `WordNetLemmatizer` 클래스를 사용합니다. 이 클래스는 `wordnet` 모듈에 포함됩니다.

In [ ]:
help(nltk.stem.wordnet)

Help on module nltk.stem.wordnet in nltk.stem:

NAME
    nltk.stem.wordnet

DESCRIPTION
    # Natural Language Toolkit: WordNet stemmer interface
    #
    # Copyright (C) 2001-2017 NLTK Project
    # Author: Steven Bird <stevenbird1@gmail.com>
    #         Edward Loper <edloper@gmail.com>
    # URL: <http://nltk.org/>
    # For license information, see LICENSE.TXT

CLASSES
    builtins.object
        WordNetLemmatizer
    
    class WordNetLemmatizer(builtins.object)
     |  WordNet Lemmatizer
     |  
     |  Lemmatize using WordNet's built-in morphy function.
     |  Returns the input word unchanged if it cannot be found in WordNet.
     |  
     |      >>> from nltk.stem import WordNetLemmatizer
     |      >>> wnl = WordNetLemmatizer()
     |      >>> print(wnl.lemmatize('dogs'))
     |      dog
     |      >>> print(wnl.lemmatize('churches'))
     |      church
     |      >>> print(wnl.lemmatize('aardwolves'))
     |      aardwolf
     |      >>> print(wnl.lemmatize('abaci'))
 

먼저 Stemming을 수행해 보겠습니다. 클래스 인스턴스를 생성하여 `stem(단어)` 메서드를 사용하는 방법입니다.

In [ ]:
from nltk.stem import LancasterStemmer

stemmer = LancasterStemmer()
print(stemmer.stem('working'), stemmer.stem('works'), stemmer.stem('worked'))
print(stemmer.stem('amusing'), stemmer.stem('amuses'), stemmer.stem('amused'))
print(stemmer.stem('happier'), stemmer.stem('happiest'))
print(stemmer.stem('fancier'), stemmer.stem('fanciest'))

work work work
amus amus amus
happy happiest
fant fanciest


- 'work': 기본 단어인 'work'에 'ing', 's', 'ed'가 붙는 단순한 변화이므로 'work'를 출력했습니다.

- 'amuse': 기본 단어인 'amuse'에서 'e'가 탈락하고 'ing', 'es', 'ed'가 붙는 복잡한 변화이므로 'amus'를 기본 단어로 인식하고 출력했습니다.

- 'happy', 'fancy': 원형을 찾지 못하고 원형에서 철자가 다른 어근 단어로 인식하고 출력했습니다.

이번에는 Lemmatization을 수행해 보겠습니다. 클래스 인스턴스를 생성하여 `lemmatize(단어, 품사)` 메서드를 사용하는 방법입니다. 품사로 전달할 인자는 동사는 `v`, 형용사는 `a`입니다.

In [ ]:
from nltk.stem import WordNetLemmatizer
import nltk
nltk.download('wordnet')

lemma = WordNetLemmatizer()
print(lemma.lemmatize('amusing', 'v'), lemma.lemmatize('amuses', 'v'), lemma.lemmatize('amused', 'v'))
print(lemma.lemmatize('happier', 'a'), lemma.lemmatize('happiest', 'a'))
print(lemma.lemmatize('fancier', 'a'), lemma.lemmatize('fanciest', 'a'))

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
amuse amuse amuse
happy happy
fancy fancy


정확하게 출력했습니다. 

# **3. Bag of Words - BOW**

**BOW** 모델은 문서가 가진 모든 단어를 문맥과 순서를 무시하고 일괄적으로 단어가 출현하는 만큼 빈도 값을 부여하여 피처 값을 추출합니다. 문서에 포함된 모든 단어를 봉투(Bag)에 넣고 섞는다는 뜻에서 Bag of Words라는 표현을 채택한 것입니다.

BOW 모델의 핵심 원리를 살펴보겠습니다. 중복된 단어를 가진 두 문장이 있다고 가정하겠습니다. 더 자세한 내용은 본서를 참고하시기 바랍니다.

1. 중복된 단어를 하나만 남기고 모두 제거합니다. 각 단어를 피처(칼럼) 형태로 나열하고 고유 인덱스를 부여합니다.

1. 각 단어가 나타난 횟수를 인덱스에 부여합니다. 예컨대 문장1과 문장2가 있다고 가정할 때, 문장1만 특정 단어가 있고 나타난 횟수가 1번이면 1을 부여하고, 문장1과 문장2 모두 특정 단어가 있고 나타난 횟수가 각각 1번과 2번이면 각각 1과 2를 부여합니다.

BOW 모델의 장점은 아래와 같습니다.

- 쉽고 빠르게 구축할 수 있습니다.

- 문서 특징을 잘 나타내며 여러 분야에서 활용합니다.

단점은 아래와 같습니다.

- 문맥 의미(Semantic Context)를 반영하기에 부족합니다. 단어 순서를 고려하지 않으므로 문맥 의미를 무시합니다. 보완책으로 n_gram 기법을 활용할 수 있지만 제한적이기 때문에 문맥 해석을 처리하기 어렵습니다.

- **희소 행렬(Sparse Matrix)** 문제가 있습니다. BOW로 피처 벡터화를 수행하면 희소 행렬 형태를 가진 데이터 세트가 만들어지기 쉽습니다. 많은 문서에서 단어를 추출하면 상당한 수의 칼럼이 생성되는데, 문서마다 서로 다른 단어로 칼럼이 구성되므로 단어가 문서마다 나타나지 않는 경우가 훨씬 많습니다. 예컨대 문서1에 단어가 수십만 개일 경우 문서2에 있는 단어는 문서1과 공통된 단어 수가 적을 가능성이 높기 때문에, 공통되지 않은 단어는 모두 0 값으로 채워집니다.

희소 행렬은 머신러닝 알고리즘의 수행 시간과 예측 성능을 떨어뜨리므로 이를 극복하기 위한 기법이 존재합니다. 아래에서 다시 다루겠습니다. 참고로 0 값이 아니라 유의미한 값으로 채워진 행렬은 **밀집 행렬(Dense Matrix)**입니다. 

## **3.1. BOW 피처 벡터화**

머신러닝 알고리즘은 숫자형 피처를 데이터로 받아서 동작하므로 텍스트 같은 데이터를 알고리즘에 바로 입력할 수 없습니다. 텍스트를 특정 의미를 가진 숫자형 값인 벡터 값으로 변환하는 **피처 벡터화** 작업을 거쳐야 합니다. 자세한 내용은 본서를 참고하시기 바랍니다. 

피처 벡터화의 원리는 아래와 같습니다. 

1. 각 문서가 가진 텍스트를 단어로 추출하여 피처로 할당합니다.

1. 각 단어의 발생 빈도와 같은 값을 피처에 값으로 부여합니다.

1. 단어 피처의 발생 빈도 값으로 구성된 벡터로 만듭니다.

즉, BOW 모델에서 피처 벡터화를 수행하는 것은 모든 문서에서 모든 단어를 칼럼 형태로 나열하고 각 문서에서 해당 단어 횟수나 정규화된 빈도를 값으로 부여하는 데이터 세트 모델로 변경하는 작업을 의미합니다. 예컨대 텍스트 문서가 M개만큼 존재하고 이 텍스트 문서들에서 모든 단어 N개를 추출하여 나열한다면 단어 N개의 값이 할당된 피처의 벡터 세트가 생성되며, 단어 피처가 $M \times N$개로 구성된 행렬이 만들어집니다.

BOW의 피처 벡터화는 아래 방식이 존재합니다.

- 카운트(Count) 기반 벡터화: 단어 피처에 값을 부여할 때 단어가 나타난 횟수(카운트)를 부여합니다. 카운트 값이 높을수록 중요한 단어입니다. 이 방식의 약점이라면 별다른 의미 없이 자주 사용될 수밖에 없는 단어까지 높은 값을 부여하기 때문에 문서 특징을 나타내는 중요한 단어에 집중하지 못합니다. 문서 개수가 적거나 텍스트 길이가 적은 경우에 적합한 방법입니다.

- **TF-IDF(Term Frequency - Inverse Document Frequency)** 기반 벡터화: 카운트 기반 벡터화의 약점을 보완하기 위한 방법입니다. 특성 문서에서 자주 나타난 단어에 높은 가중치를 부여하지만 모든 문서에서 전반적으로 자주 나타나는 단어에는 페널티(Penalty)를 부여합니다. 예컨대 언어 특성상 모든 문서에서 보편적이고 반복적으로 사용될 수 있는 부사('많이', '상당하게', 당연하게' 등)나 명사('이름', '업무', '조직' 등)와 같은 단어들은 중요하지 않으므로(물론 형용사도 포함될 수 있습니다) 페널티를 부여하여 가중치 균형을 잡습니다. 문서 개수가 많거나 텍스트 길이가 많은 경우에 적합한 방법입니다. 

## **3.2. 사이킷런의 Count 및 TF-IDF 벡터화 구현: CountVectorizer, TfidfVectorizer**

`CounterVectorizer` 클래스는 카운트 기반 벡터화를 구현한 클래스이며 소문자 일괄 변환, 토큰화, 스톱 워드 필터링 등 전처리도 수행합니다. 파라미터는 아래와 같습니다. 참고로 TF-IDF를 구현한 `TfidfVectorizer` 클래스는 `CountVectorizer` 클래스의 파라미터와 변환 방법이 동일합니다.


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

help(CountVectorizer)

Help on class CountVectorizer in module sklearn.feature_extraction.text:

class CountVectorizer(_VectorizerMixin, sklearn.base.BaseEstimator)
 |  CountVectorizer(*, input='content', encoding='utf-8', decode_error='strict', strip_accents=None, lowercase=True, preprocessor=None, tokenizer=None, stop_words=None, token_pattern='(?u)\\b\\w\\w+\\b', ngram_range=(1, 1), analyzer='word', max_df=1.0, min_df=1, max_features=None, vocabulary=None, binary=False, dtype=<class 'numpy.int64'>)
 |  
 |  Convert a collection of text documents to a matrix of token counts.
 |  
 |  This implementation produces a sparse representation of the counts using
 |  scipy.sparse.csr_matrix.
 |  
 |  If you do not provide an a-priori dictionary and you do not use an analyzer
 |  that does some kind of feature selection then the number of features will
 |  be equal to the vocabulary size found by analyzing the data.
 |  
 |  Read more in the :ref:`User Guide <text_feature_extraction>`.
 |  
 |  Parameters
 |  -----

`CountVectorizer` 클래스를 활용하여 카운트 기반 피처 벡터화를 수행하는 방법은 아래와 같습니다.

1. 데이터 전처리: 영어의 경우 단어를 소문자로 변경하는 등 전처리 작업을 수행합니다. 디폴트는 `lowercase=Ture`입니다.

1. 디폴트 `analyzer=word`인 단어 기준으로 `n_gram_range` 파라미터를 반영하여 단어를 토큰화합니다. `n_gram_range` 파라미터는 단어 순서를 보강하기 위한 범위를 지정하는 것이며 `(1, 1)`로 지정하면 토큰화된 단어를 1개씩 피처로 추출하고 `(1, 2)`로 지정하면 토큰화된 단어를 최소 1개씩 순서대로 최대 2개씩(최대 2개) 묶어서 추출합니다. 

1. 텍스트 정규화를 수행합니다. 단, `stop_words='english'`로 파라미터를 지정하면 스톱 워드만 필터링 합니다. Stemming이나 Lemmatization 같은 어근 변환은 `CountVectorizer` 클래스에서 지원하진 않지만 `tokenizer` 파라미터에 어근 변환 함수를 따로 선언하여 적용하거나 외부 패키지를 적용하면 어근 변환을 수행할 수 있습니다.

1. `max_df`, `min_df`, `max_features` 등 파라미터를 이용하여 토큰화된 단어를 피처로 추출하고 단어 빈도수 벡터 값을 적용합니다. 

 - `max_df`: 높은 빈도수를 가진 단어 피처를 제외합니다. 자주 나타나는 단어는 스톱 워드와 비슷한 반복적인 단어일 가능성이 높습니다.

 - `min_df`: 낮은 빈도수를 가진 단어 피처를 제외합니다. 적게 나타나는 단어는 중요하지 않은 단어일 가능성이 높습니다.

 - `max_features`: 추출할 피처 개수를 제한합니다.

## **3.3. BOW 벡터화를 위한 희소 행렬**

대규모의 행렬이 생성될 때 레코드의 각 문서가 가지는 단어 수는 제한적이므로 행렬 값은 대부분 0입니다. 이렇게 대규모 행렬의 값이 대부분 0인 행렬을 희소 행렬이라고 합니다. BOW 형태를 가진 언어 모델의 피처 벡터화는 대부분 희소 행렬입니다.

다만 희소 행렬은 불필요한 0 값이 메모리 공간에 지나치게 많이 할당되어 많은 메모리 공간이 필요하고 행렬 크기가 커서 연산할 때 데이터 엑세스를 위한 오랜 시간이 소요됩니다. 물리적으로 적은 메모리 공간을 사용하도록 변환하는 방법은 **COO** 형식과 **CSR** 형식이 있습니다. 이 중에 CSR이 더 뛰어나고 많이 사용됩니다.

## **3.4. 희소 행렬 - COO 형식**

**COO(Coordinate: 좌표)** 형식은 0이 아닌 데이터만 별도 데이터 배열에 저장하고, 그 데이터가 가리키는 행과 열 위치를 별도 배열로 저장합니다. 예컨대 아래 2차원 데이터인 밀집 행렬이 있다고 가정하겠습니다.

```
[[3, 0, 1],
 [0, 2, 0]]
```

여기서 0이 아닌 데이터 `[3, 1, 2]`를 위치(row, col)로 표시하면 `(0, 0), (0, 2), (1, 1)`입니다. 로우(row)와 칼럼(col)을 따로 떼어 별도 배열로 저장하면 로우는 `[0, 0, 1]`이고 칼럼은 `[0, 2, 1]`입니다.

사이파이의 `sparse` 패키지를 사용하면 희소 행렬 변환을 수행할 수 있습니다. 위 2차원 데이터를 COO 형식으로 희소 행렬로 변환해 보겠습니다. 사용할 클래스는 `coo_matrix`입니다. 참고로 `coo_matrix`의 명칭은 소문자로 시작하지만 메서드가 아닌 클래스입니다.

In [ ]:
import numpy as np
from scipy import sparse

# 밀집 행렬을 생성합니다.
dense = np.array([[3, 0, 1], [0, 2, 0]])

# 0이 아닌 데이터를 추출합니다.
data = np.array([3, 1, 2])

# 행 위치와 열 위치를 배열로 생성합니다.
row_pos = np.array([0, 0, 1])
col_pos = np.array([0, 2, 1])

# COO 형식으로 희소 행렬을 생성하고 인스턴스화합니다.
sparse_coo = sparse.coo_matrix((data, (row_pos, col_pos)))

# `toarray()` 메서드를 사용하여 밀집 행렬로 만듭니다.
sparse_coo.toarray()

array([[3, 0, 1],
       [0, 2, 0]])

COO 형식의 문제점이라면 행과 열 위치를 나타내기 위해 반복적으로 위치 데이터를 사용해야 하는 것입니다. 예컨대 아래 2차원 배열이 있다고 가정하겠습니다.

```
[[0, 0, 1, 0, 0, 5],
 [1, 4, 0, 3, 2, 5],
 [0, 6, 0, 3, 0, 0],
 [2, 0, 0, 0, 0, 0],
 [0, 0, 0, 7, 0, 8],
 [1, 0, 0, 0, 0, 0]]
```

여기서 0이 아닌 데이터 배열은 아래와 같습니다.

`[1, 5, 1, 4, 3, 2, 5, 6, 3, 2, 7, 8, 1]`

행 위치 배열은 아래와 같습니다.

`[0, 0, 1, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5]`

열 위치 배열은 아래와 같습니다.

`[2, 5, 0, 1, 3, 4, 5, 1, 3, 0, 3, 5, 0]`

이 중에 행 위치 배열은 같은 값이 반복적으로 나열되고 순차적으로 증가합니다. 0은 2번, 1은 5번, 2는 2번인 식입니다. 따라서 행 위치 배열의 고유한 값의 시작 위치만 표기한다면(위치의 위치를 표기한다면) 불필요한 반복을 제거할 수 있을 것입니다. 인덱스를 기준으로 삼아 0의 시작 위치는 0, 1의 시작 위치는 2, 2의 시작 위치는 7, 3의 시작 위치는 9, 4의 시작 위치는 10, 5의 시작 위치는 12이므로 `[0, 2, 7, 9, 10, 12]`가 됩니다. 최종적으로 이 배열의 맨 마지막에 데이터의 총 항목 개수를 추가하면 `[0, 2, 7, 9, 10, 12, 13]`이 됩니다. 이 일련의 방식이 바로 아래에서 기술할 **CSR(Compressed Sparse Row)** 형식입니다.

## **3.5. 희소 행렬 - CSR 형식**

**CSR(Compressed Sparse Row)** 형식은 COO 형식보다 메모리가 적게 필요하고 빠르게 연산할 수 있습니다. `csr_matrix` 클래스에 구현되어 있으며 전달할 인자는 `(0이 아닌 데이터 배열, 열 위치 배열, 행 위치 배열의 고유한 값의 시작 위치)`입니다.

CSR 형식을 코드로 구현해 보겠습니다.

In [ ]:
dense2 = np.array([[0, 0, 1, 0, 0, 5],
                   [1, 4, 0, 3, 2, 5],
                   [0, 6, 0, 3, 0, 0],
                   [2, 0, 0, 0, 0, 0],
                   [0, 0, 0, 7, 0, 8],
                   [1, 0, 0, 0, 0, 0]])

data2 = np.array([1, 5, 1, 4, 3, 2, 5, 6, 3, 2, 7, 8, 1])
row_pos = np.array([0, 0, 1, 1, 1, 1, 1, 2, 2, 3, 4, 4, 5])
col_pos = np.array([2, 5, 0, 1, 3, 4, 5, 1, 3, 0, 3, 5, 0])
row_pos_ind = np.array([0, 2, 7, 9, 10, 12, 13])

sparse_csr = sparse.csr_matrix((data2, col_pos, row_pos_ind))
print(sparse_csr.toarray())

[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


실제로 사용할 때는 생성 파라미터로 밀집 행렬만 전달하면 됩니다.

In [ ]:
dense3 = np.array([[0, 0, 1, 0, 0, 5],
                   [1, 4, 0, 3, 2, 5],
                   [0, 6, 0, 3, 0, 0],
                   [2, 0, 0, 0, 0, 0],
                   [0, 0, 0, 7, 0, 8],
                   [1, 0, 0, 0, 0, 0]])

coo = sparse.coo_matrix(dense3)
csr = sparse.csr_matrix(dense3)
print(coo.toarray())
print()
print(csr.toarray())

[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]

[[0 0 1 0 0 5]
 [1 4 0 3 2 5]
 [0 6 0 3 0 0]
 [2 0 0 0 0 0]
 [0 0 0 7 0 8]
 [1 0 0 0 0 0]]


참고로 `CountVectorizer` 클래스와 `TfidfVectorizer` 클래스로 변환된 피처 벡터화 행렬은 모두 CSR 형태의 희소 행렬입니다.

# **4. 텍스트 분류 실습 - 20 뉴스그룹 분류**

사이킷런의 20 뉴스그룹 데이터 세트를 사용하여 텍스트 분류를 수행해 보겠습니다. 텍스트 분류는 특정 문서의 분류를 학습 데이터를 토대로 학습하고 모델을 생성하여 다른 문서의 분류를 예측하는 것입니다.

텍스트를 피처 벡터화로 변환하면 일반적으로 희소 행렬이 됩니다. 이 희소 행렬에 분류를 효과적으로 처리할 수 있는 알고리즘은 **로지스틱 회귀**, **선형 서포트 벡터 머신**, **나이브 베이즈** 등입니다. 텍스트를 기반으로 분류를 수행하려면 먼저 텍스트를 정규화하고 피처 벡터화를 적용해야 합니다. 그리고나서 알고리즘을 적용하고 분류를 학습하여 예측하고 평가합니다. 

이번 파트에서는 카운트 기반과 TF-IDF 기반 벡터화를 모두 적용하여 예측 성능을 비교하고 피처 백터화를 위한 파라미터와 `GridSearchCV` 기반의 하이퍼 파라미터 튜닝, `Pipeline` 인스턴스를 통한 피처 벡터화 파라미터와 `GridSearchCV` 기반의 하이퍼 파라미터 튜닝을 전부 수행해 보겠습니다.

## **4.1. 텍스트 정규화**

`fetch_20newsgroups()` 메서드를 호출하여 데이터를 준비하겠습니다.

In [ ]:
from sklearn.datasets import fetch_20newsgroups

news_data = fetch_20newsgroups(subset='all', random_state=156)

In [ ]:
help(fetch_20newsgroups)

Help on function fetch_20newsgroups in module sklearn.datasets._twenty_newsgroups:

fetch_20newsgroups(*, data_home=None, subset='train', categories=None, shuffle=True, random_state=42, remove=(), download_if_missing=True, return_X_y=False)
    Load the filenames and data from the 20 newsgroups dataset (classification).
    
    Download it if necessary.
    
    =================   ==========
    Classes                     20
    Samples total            18846
    Dimensionality               1
    Features                  text
    =================   ==========
    
    Read more in the :ref:`User Guide <20newsgroups_dataset>`.
    
    Parameters
    ----------
    data_home : str, default=None
        Specify a download and cache folder for the datasets. If None,
        all scikit-learn data is stored in '~/scikit_learn_data' subfolders.
    
    subset : {'train', 'test', 'all'}, default='train'
        Select the dataset to load: 'train' for the training set, 'test'
        fo

데이터 구성을 살펴보겠습니다. 자세한 설명은 생략합니다.

In [ ]:
print(news_data.keys())

dict_keys(['data', 'filenames', 'target_names', 'target', 'DESCR'])


In [ ]:
print(news_data.DESCR)

.. _20newsgroups_dataset:

The 20 newsgroups text dataset
------------------------------

The 20 newsgroups dataset comprises around 18000 newsgroups posts on
20 topics split in two subsets: one for training (or development)
and the other one for testing (or for performance evaluation). The split
between the train and test set is based upon a messages posted before
and after a specific date.

This module contains two loaders. The first one,
:func:`sklearn.datasets.fetch_20newsgroups`,
returns a list of the raw texts that can be fed to text feature
extractors such as :class:`~sklearn.feature_extraction.text.CountVectorizer`
with custom parameters so as to extract feature vectors.
The second one, :func:`sklearn.datasets.fetch_20newsgroups_vectorized`,
returns ready-to-use features, i.e., it is not necessary to use a feature
extractor.

**Data Set Characteristics:**

    =================   ==========
    Classes                     20
    Samples total            18846
    Dimensionality

먼저 타겟 클래스를 살펴보겠습니다.

In [ ]:
import pandas as pd

print('target 클래스의 값과 분포도\n', pd.Series(news_data.target).value_counts().sort_index())
print()
print('target 클래스의 이름\n', news_data.target_names)
print(len(news_data.target_names))

target 클래스의 값과 분포도
 0     799
1     973
2     985
3     982
4     963
5     988
6     975
7     990
8     996
9     994
10    999
11    991
12    984
13    990
14    987
15    997
16    910
17    940
18    775
19    628
dtype: int64

target 클래스의 이름
 ['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']
20


총 20개의 타겟 클래스를 가졌으며 각 클래스의 값 분포는 몇 클래스를 제외하면 고르게 분포합니다.

데이터를 하나만 살펴보겠습니다.

In [ ]:
print(news_data.data[0])

From: egreen@east.sun.com (Ed Green - Pixel Cruncher)
Subject: Re: Observation re: helmets
Organization: Sun Microsystems, RTP, NC
Lines: 21
Distribution: world
Reply-To: egreen@east.sun.com
NNTP-Posting-Host: laser.east.sun.com

In article 211353@mavenry.altcit.eskimo.com, maven@mavenry.altcit.eskimo.com (Norman Hamer) writes:
> 
> The question for the day is re: passenger helmets, if you don't know for 
>certain who's gonna ride with you (like say you meet them at a .... church 
>meeting, yeah, that's the ticket)... What are some guidelines? Should I just 
>pick up another shoei in my size to have a backup helmet (XL), or should I 
>maybe get an inexpensive one of a smaller size to accomodate my likely 
>passenger? 

If your primary concern is protecting the passenger in the event of a
crash, have him or her fitted for a helmet that is their size.  If your
primary concern is complying with stupid helmet laws, carry a real big
spare (you can put a big or small head in a big helmet, bu

기사 내용뿐만 아니라 제목, 작성자, 소속, 이메일 등 정보를 모두 갖고 있습니다. 모든 정보를 사용하면 어떤 머신러닝 알고리즘을 적용해도 높은 예측 성능을 발휘하게 되므로 내용을 제외한 모든 정보를 삭제하겠습니다.

In [ ]:
from sklearn.datasets import fetch_20newsgroups

train_news = fetch_20newsgroups(subset='train', remove=('headers', 'footers', 'quotes'),
                   random_state=156)
X_train = train_news.data
y_train = train_news.target

test_news = fetch_20newsgroups(subset='test', remove=('headers', 'footers', 'quotes'),
                               random_state=156)
X_test = test_news.data
y_test = test_news.target

print('학습 데이터 크기:', len(train_news.data))
print('테스트 데이터 크기:', len(test_news.data))

학습 데이터 크기: 11314
테스트 데이터 크기: 7532


In [ ]:
print(train_news.data[0])
print()
print(test_news.data[0])



What I did NOT get with my drive (CD300i) is the System Install CD you
listed as #1.  Any ideas about how I can get one?  I bought my IIvx 8/120
from Direct Express in Chicago (no complaints at all -- good price & good
service).

BTW, I've heard that the System Install CD can be used to boot the mac;
however, my drive will NOT accept a CD caddy is the machine is off.  How can
you boot with it then?

--Dave


The tech support line for GCC is 1-800-231-1570.


## **4.2. 피처 벡터화 변환과 머신러닝 모델 학습/예측/평가**

`CountVectorizer` 클래스를 사용하여 학습 데이터와 테스트 데이터의 텍스트를 피처 벡터화하겠습니다.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

cnt_vect = CountVectorizer()
cnt_vect.fit(X_train)
X_train_cnt_vect = cnt_vect.transform(X_train)

# 학습 데이터와 테스트 데이터의 피처 개수를 같게 만들기 위해
# 학습 데이터로 `fit()`을 적용한 인스턴스를 사용하여 테스트 데이터를 변환합니다.
X_test_cnt_vect = cnt_vect.transform(X_test)

print('학습 데이터 텍스트의 CountVectorizer Shape:', X_train_cnt_vect.shape)
print('테스트 데이터 텍스트의 CountVectorizer Shape:', X_test_cnt_vect.shape)

학습 데이터 텍스트의 CountVectorizer Shape: (11314, 101631)
테스트 데이터 텍스트의 CountVectorizer Shape: (7532, 101631)


두 데이터 세트에 각각 같은 단어수인 101,631개가 피처로 생성되었습니다.

로지스틱 회귀를 적용하여 뉴스그룹을 예측해 보겠습니다.

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

lr_clf = LogisticRegression()
lr_clf.fit(X_train_cnt_vect, y_train)
pred = lr_clf.predict(X_test_cnt_vect)

print('CountVectorized Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

CountVectorized Logistic Regression의 예측 정확도: 0.6076739245884227


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


이번에는 `TfidfVectorizer` 클래스를 사용하여 피처 벡터화를 수행하고 로지스틱 회귀를 적용해 보겠습니다.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vect = TfidfVectorizer()
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

lr_clf = LogisticRegression()
lr_clf.fit(X_train_tfidf_vect, y_train)
pred = lr_clf.predict(X_test_tfidf_vect)

print('TF-IDF Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

TF-IDF Logistic Regression의 예측 정확도: 0.6736590546999469


TF-IDF 기반 예측 정확도가 더 높습니다. 다량의 문서와 텍스트는 TF-IDF 기반 피처 벡터화가 더 좋은 예측 결과를 도출합니다.

TA를 수행할 때 머신러닝 모델 성능을 제고하려면 최적 알고리즘을 선택하고 피처 전처리를 완벽하게 수행해야 합니다. 이번에는 다양한 파라미터를 지정하여 예측 성능을 제고해 보겠습니다.

In [ ]:
tfidf_vect = TfidfVectorizer(stop_words='english', ngram_range=(1, 2), max_df=300)
tfidf_vect.fit(X_train)
X_train_tfidf_vect = tfidf_vect.transform(X_train)
X_test_tfidf_vect = tfidf_vect.transform(X_test)

lr_clf = LogisticRegression()
lr_clf.fit(X_train_tfidf_vect, y_train)
pred = lr_clf.predict(X_test_tfidf_vect)

print('TF-IDF Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

TF-IDF Logistic Regression의 예측 정확도: 0.6922464152947424


성능이 약간 향상되었습니다.

그리드 서치를 활용하여 로지스틱 회귀의 최적 파라미터를 탐색(`C`만 탐색 시도)한 뒤 모델의 예측 성능을 확인해 보겠습니다.

In [ ]:
from sklearn.model_selection import GridSearchCV

params = {'C': [0.01, 0.1, 1, 5, 10]}
grid_cv_lr = GridSearchCV(lr_clf, param_grid=params, cv=3, scoring='accuracy', verbose=1)
grid_cv_lr.fit(X_train_tfidf_vect, y_train)
print('Logistic Regression best C parameter:', grid_cv_lr.best_params_)

Fitting 3 folds for each of 5 candidates, totalling 15 fits


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


Logistic Regression best C parameter: {'C': 10}


/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


In [ ]:
pred = grid_cv_lr.predict(X_test_tfidf_vect)
print('TF-IDF Vectorized Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

TF-IDF Vectorized Logistic Regression의 예측 정확도: 0.7010090281465746


성능이 약간 더 향상되었습니다.

## **4.3. 사이킷런 파이프라인(Pipeline) 사용 및 GridSearchCV와의 결합**

위 과정을 파이프라인으로 구현해 보겠습니다. 먼저 그리드서치를 결합하지 않은 형태를 구현하겠습니다.

In [ ]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
                     ('tfidf_vect', TfidfVectorizer(stop_words='english', ngram_range=(1, 2), max_df=300)),
                     ('lr_clf', LogisticRegression(C=10))
                     ])
pipeline.fit(X_train, y_train)
pred = pipeline.predict(X_test)

print('Pipeline을 통한 Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


Pipeline을 통한 Logistic Regression의 예측 정확도: 0.7010090281465746


같은 결괏값이 도출되었습니다.

이번에는 파이프라인과 그리드서치를 결합해 보겠습니다. 관련된 여러 파라미터를 추가하여 수행하되 모든 파라미터를 탐색하면 많은 수행 시간이 소요되므로 몇 개만 선별하여 시도하겠습니다.


In [ ]:
!nvidia-smi

Sat May 21 09:33:45 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla K80           Off  | 00000000:00:04.0 Off |                    0 |
| N/A   50C    P8    31W / 149W |      0MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
                     ('tfidf_vect', TfidfVectorizer(stop_words='english')),
                     ('lr_clf', LogisticRegression())
                     ])
# 언더바 2개(`__`)를 이어서 탐색할 파라미터를 특정합니다.
params = {'tfidf_vect__ngram_range': [(1, 1), (1, 2), (1, 3)],
          'tfidf_vect__max_df': [100, 300, 700],
          'lr_clf__C': [1, 5, 10]
          }
grid_cv_pipe = GridSearchCV(pipeline, param_grid=params, cv=3, scoring='accuracy', verbose=1)
grid_cv_pipe.fit(X_train, y_train)
print(grid_cv_pipe.best_params_, grid_cv_pipe.best_score_)

Fitting 3 folds for each of 27 candidates, totalling 81 fits


코랩 무료 버전으로 코드를 실행하면 도중에 램 부족 문제가 발생하여 세션이 다운됩니다. 램 12.69GB을 모두 사용할 만큼 무거운 작업인가 봅니다. 보유 중인 랩톱으로도 시도했으나 수행 시간이 10시간 이상 걸려서 연산을 중지하기로 결정했습니다. 본서에 명시된 출력 결과를 명시하는 것으로 갈음하겠습니다.

```
[Parallel(n_jobs=1)]: Done 81 out of 81 | elapsed: 31.9min finished
{'lr_clf__C': 10, 'tfidf_vect__max_df': 700, 'tfidf_vect__ngram_range': (1, 2)} 0.755524129397207
```

예측 정확도를 확인해 보겠습니다.

In [ ]:
pred = grid_cv_pipe.predict(X_test)
print('Pipeline을 통한 Logistic Regression의 예측 정확도:', accuracy_score(y_test, pred))

```
Pipeline을 통한 Logistic Regression의 예측 정확도: 0.702
```

이전 결과에 비하면 거의 개선되지 않았다고 볼 수 있습니다. 서포트 벡터머신과 나이브 베이즈 알고리즘도 시도해볼 가치가 있겠습니다.

# **5. 감성 분석**

## **5.1. 감성 분석 소개**

**감성 분석(Sentiment Analysis)**은 문서의 주관적인 감성, 의견, 감정, 기분 등을 파악하기 위한 방법입니다. 소셜 미디어, 여론조사, 온라인 리뷰, 피드백 등 다양한 분야에서 활용됩니다. 문서 내 텍스트가 나타내는 주관적인 단어와 문맥을 기반으로 감성 수치(긍정 감성 지수, 부정 감성 지수)를 계산하고 도출하여 긍정과 부정을 결정합니다.

감성 분석은 지도학습과 비지도학습으로 나눌 수 있습니다. 지도학습은 학습 데이터와 타깃 레이블 값을 기반으로 감성 분석 학습을 수행하고 다른 데이터의 감성 분석을 예측합니다. 텍스트 기반 분류와 동일하다고 볼 수 있습니다. 비지도학습은 `Lexicon`이라는 감성 어휘 사전 모듈을 사용합니다. 감성 분석을 위한 용어와 문맥에 대한 정보를 가진 이 모듈로 긍정과 부정을 판단합니다.


## **5.2. 지도학습 기반 감성 분석 실습 - IMDB 영화평**

지도학습을 기반으로 감성 분석을 수행해 보겠습니다. [IMDB 영화평](https://www.kaggle.com/c/word2vec-nlp-tutorial/data) 데이터를 사용하여 긍정과 부정 여부를 예측하는 모델을 생성하겠습니다.

데이터를 준비하겠습니다.

In [2]:
from google.colab import drive
drive.mount('/content/gdrive/')
import pandas as pd

review_df = pd.read_csv('/content/gdrive/MyDrive/Bag of Words Meets Bags of Popcorn/labeledTrainData.tsv',
                        header=0, sep='\t', quoting=3)
review_df.head()

Mounted at /content/gdrive/


id  sentiment                                             review
0  "5814_8"          1  "With all this stuff going down at the moment ...
1  "2381_9"          1  "\"The Classic War of the Worlds\" by Timothy ...
2  "7759_3"          0  "The film starts with a manager (Nicholas Bell...
3  "3630_4"          0  "It must be assumed that those who praised thi...
4  "9495_8"          1  "Superbly trashy and wondrously unpretentious ...

리뷰가 어떻게 구성되어 있는지 확인해 보겠습니다.

In [ ]:
print(review_df['review'][0])

"With all this stuff going down at the moment with MJ i've started listening to his music, watching the odd documentary here and there, watched The Wiz and watched Moonwalker again. Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent. Moonwalker is part biography, part feature film which i remember going to see at the cinema when it was originally released. Some of it has subtle messages about MJ's feeling towards the press and also the obvious message of drugs are bad m'kay.<br /><br />Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring. Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him.<br /><br />The actual feature film bit when it finally sta

`<br />` 태그가 있습니다. 불필요하므로 모두 공백으로 바꾸겠습니다.

In [3]:
review_df['review'] = review_df['review'].str.replace('<br />', ' ')

아울러 숫자나 특수문자도 감성 분석에는 무의미한 요소입니다. 공백으로 대체하겠습니다. `re` 모듈은 편리하게 정규 표현식을 지원하며 텍스트 분석을 위한 필수 도구입니다.

In [ ]:
help(re)

Help on module re:

NAME
    re - Support for regular expressions (RE).

MODULE REFERENCE
    https://docs.python.org/3.7/library/re
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    This module provides regular expression matching operations similar to
    those found in Perl.  It supports both 8-bit and Unicode strings; both
    the pattern and the strings being processed can contain null bytes and
    characters outside the US ASCII range.
    
    Regular expressions can contain both special and ordinary characters.
    Most ordinary characters, like "A", "a", or "0", are the simplest
    regular expressions; they simply match themselves.  You can
    concatenate ordinary characters, so last mat

In [4]:
import re

# `re`모듈의 `sub()` 메서드를 적용하여 숫자와 특수문자를 공백으로 변환합니다.
# '[^a-zA-Z]'는 영어 대소문자가 아닌 모든 문자를 찾는 것을 의미합니다.
review_df['review'] = review_df['review'].apply(lambda x: re.sub('[^a-zA-Z]', ' ', x))

In [ ]:
print(review_df['review'][0])

 With all this stuff going down at the moment with MJ i ve started listening to his music  watching the odd documentary here and there  watched The Wiz and watched Moonwalker again  Maybe i just want to get a certain insight into this guy who i thought was really cool in the eighties just to maybe make up my mind whether he is guilty or innocent  Moonwalker is part biography  part feature film which i remember going to see at the cinema when it was originally released  Some of it has subtle messages about MJ s feeling towards the press and also the obvious message of drugs are bad m kay   Visually impressive but of course this is all about Michael Jackson so unless you remotely like MJ in anyway then you are going to hate this and find it boring  Some may call MJ an egotist for consenting to the making of this movie BUT MJ and most of his fans would say that he made it for the fans which if true is really nice of him   The actual feature film bit when it finally starts is only on for  

'review' 피처를 피처 데이터 세트로 생성하고, 'sentiment' 피처를 결정 값 데이터 세트로 생성하겠습니다. 이를 토대로 학습 데이터 세트와 테스트 데이터 세트로 분할하겠습니다.

In [5]:
from sklearn.model_selection import train_test_split

feature_df = review_df.drop(['id', 'sentiment'], axis=1, inplace=False)
class_df = review_df['sentiment']

X_train, X_test, y_train, y_test = train_test_split(feature_df, class_df,
                                                    test_size=0.3, random_state=156)
print(X_train.shape, X_test.shape)
print(y_train.shape, y_test.shape)

(17500, 1) (7500, 1)
(17500,) (7500,)


리뷰 텍스트를 카운트 기반으로 피처 벡터화를 수행하고 머신러닝 알고리즘을 적용하여 예측 성능을 확인하겠습니다.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, roc_auc_score

pipeline = Pipeline([
                     ('cnt_vect', CountVectorizer(stop_words='english', ngram_range=(1, 2))),
                     ('lr_clf', LogisticRegression(C=10))
])
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:, 1]

print('예측 정확도:', accuracy_score(y_test, pred))
print('ROC-AUC:', roc_auc_score(y_test, pred_probs))

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,


예측 정확도: 0.886
ROC-AUC: 0.9502705298201685


이번엔 TF-IDF 기반으로 피처 벡터화를 수행하고 같은 과정을 구현해 보겠습니다.

In [ ]:
pipeline = Pipeline([
                     ('tfidf_vect', TfidfVectorizer(stop_words='english', ngram_range=(1, 2))),
                     ('lr_clf', LogisticRegression(C=10))
])
pipeline.fit(X_train['review'], y_train)
pred = pipeline.predict(X_test['review'])
pred_probs = pipeline.predict_proba(X_test['review'])[:, 1]

print('예측 정확도:', accuracy_score(y_test, pred))
print('ROC-AUC:', roc_auc_score(y_test, pred_probs))

예측 정확도: 0.8936
ROC-AUC: 0.959799823582973


성능이 약간 향상되었습니다.

## **5.3. 비지도학습 기반 감성 분석 소개**

일반적으로 감성 분석용 데이터는 결정된 레이블 값을 갖고 있지 않으므로 비지도 감성 분석은 **Lexicon(어휘집)**을 기반으로 수행합니다. 참고로 한글은 Lexicon이 없습니다.

비지도 학습에서 Lexicon은 주로 감성만을 분석하기 위해 지원하는 감성 어휘 사전입니다. 이 감성 사전은 긍정 감성이나 부정 감성의 정도를 나타내는 수치인 **감성 지수(Polarity score)**를 갖고 있습니다. 감성 지수는 단어 위치나 주변 단어, 문맥, POS(Part of Speech, 품사) 등을 고려하여 결정됩니다. 감성 사전을 구현한 패키지는 `NLTK`이며 많은 서브 모듈이 포함되어 있고, 감성 사전을 구현한 모듈은 `Lexicon`입니다.

NLP 패키지의 `WordNet` 모듈은 영어 어휘 사전이며 시맨틱(Semantic, 문맥상 의미) 분석을 제공합니다. 같은 어휘라도 달리 사용되는 시맨틱 정보를 제공하며, 여러 품사로 구성된 개별 단어를 **Synset(Sets of cognitive synonyms)**이라는 개념을 토대로 표현합니다. Synset은 단순한 단어 하나가 아닌 그 단어가 가진 시맨틱 정보를 제공하는 개념입니다.

다만 `NLTK` 패키지의 감성 사전은 좋지 않은 예측 성능을 보이기 때문에 실무에서는 `NLTK` 패키지가 아닌 다른 감성 사전을 적용합니다. 아래는 `NLTK`를 포함한 대표적인 감성 사전 몇 가지 목록입니다.

- `SentiWordNet`: 감성 단어 전용의 `WordNet`을 구현했으며, `WordNet`의 Synset 개념을 감성 분석에 적용했습니다(`sentiwordnet` 인스턴스로 구현되어 있습니다). 이 Synset별로 3가지 감성 점수(sentiment score)인 긍정 감성 지수, 부정 감성 지수, 객관성 지수를 할당합니다. 문장별 단어들의 긍정 감성 지수와 부정 감성 지수를 합산하여 최종 감성 지수를 계산하고 긍정과 부정 여부를 판단합니다.

- `VADER`: 소셜 미디어의 텍스트에 대한 감성 분석을 제공하는 패키지입니다. 뛰어난 감성 분석 결과를 제공하고 빠른 수행 시간을 보장하여 대용량 텍스트 데이터에 유용합니다.

- `Pattern` 예측 성능 측면에서 가장 주목받는 패키지입니다. 다만 파이썬 3 버전에서는 호환이 되지 않습니다. 

## **5.4. SentiWordNet을 이용한 감성 분석**

`SentiWordNet`은 예측 정확도가 높지 않아서 잘 사용되지는 않습니다. 다만 감성 사전의 구성 방식과 시맨틱 기반 사전 구축 방식을 이해할 목적이라면 한 번쯤 살펴볼 가치가 있습니다. 실사용은 `sentiwordnet` 인스턴스를 호출하여 사용합니다. 이 인스턴스는 `LazyCorpusLoader` 클래스의 인스턴스입니다. 

In [ ]:
from nltk.corpus import sentiwordnet as swn

help(swn)

Help on LazyCorpusLoader in module nltk.corpus.util object:

sentiwordnet = class LazyCorpusLoader(builtins.object)
 |  sentiwordnet(name, reader_cls, *args, **kwargs)
 |  
 |  To see the API documentation for this lazily loaded corpus, first
 |  run corpus.ensure_loaded(), and then run help(this_corpus).
 |  
 |  LazyCorpusLoader is a proxy object which is used to stand in for a
 |  corpus object before the corpus is loaded.  This allows NLTK to
 |  create an object for each corpus, but defer the costs associated
 |  with loading those corpora until the first time that they're
 |  actually accessed.
 |  
 |  The first time this object is accessed in any way, it will load
 |  the corresponding corpus, and transform itself into that corpus
 |  (by modifying its own ``__class__`` and ``__dict__`` attributes).
 |  
 |  If the corpus can not be found, then accessing this object will
 |  raise an exception, displaying installation instructions for the
 |  NLTK data package.  Once they've pr

### **5.4.1. WordNet Synset과 SentiWordNet SentiSynset 클래스의 이해**

`SentiwordNet`은 `WordNet` 기반의 Synset을 이용하므로 Synset 개념부터 살피겠습니다. `nltk` 패키지의 모든 데이터 세트와 패키지를 다운로드하겠습니다.


In [ ]:
import nltk
nltk.download('all')

`wordnet` 모듈을 토대로 `synsets()` 메서드를 사용하여 단어의 Synset을 추출하겠습니다. 

In [ ]:
from nltk.corpus import wordnet as wn

term = 'present'
synsets = wn.synsets(term)
print('synsets type:', type(synsets))
print('synsets 값 개수:', len(synsets))
print('synsets 값 :', synsets)

synsets type: <class 'list'>
synsets 값 개수: 18
synsets 값 : [Synset('present.n.01'), Synset('present.n.02'), Synset('present.n.03'), Synset('show.v.01'), Synset('present.v.02'), Synset('stage.v.01'), Synset('present.v.04'), Synset('present.v.05'), Synset('award.v.01'), Synset('give.v.08'), Synset('deliver.v.01'), Synset('introduce.v.01'), Synset('portray.v.04'), Synset('confront.v.03'), Synset('present.v.12'), Synset('salute.v.06'), Synset('present.a.01'), Synset('present.a.02')]


`Synset` 클래스를 리스트 형태로 반환했슨비다. `synsets` 객체의 값 중에 `Synset('present.n.01')`만 살펴보겠습니다. 이 값은 POS(품사) 태그를 의미하며, 'present'는 의미, 'n'은 품사(명사를 의미함), '01'은 'present'가 명사로서 가지는 여러 의미를 구분하는 인덱스입니다.

`synsets` 객체의 속성을 확인해 보겠습니다. Synset은 POS, 정의(Definition), 부명제(Lemma) 등으로 시맨틱적인 요소를 표현할 수 있습니다.

In [ ]:
for synset in synsets:
    print('- Synset name:', synset.name(), '-')
    print('POS:', synset.lexname())
    print('Definition:', synset.definition())
    print('Lemmas:', synset.lemma_names())
    print()

- Synset name: present.n.01 -
POS: noun.time
Definition: the period of time that is happening now; any continuous stretch of time including the moment of speech
Lemmas: ['present', 'nowadays']

- Synset name: present.n.02 -
POS: noun.possession
Definition: something presented as a gift
Lemmas: ['present']

- Synset name: present.n.03 -
POS: noun.communication
Definition: a verb tense that expresses actions or states at the time of speaking
Lemmas: ['present', 'present_tense']

- Synset name: show.v.01 -
POS: verb.perception
Definition: give an exhibition of to an interested audience
Lemmas: ['show', 'demo', 'exhibit', 'present', 'demonstrate']

- Synset name: present.v.02 -
POS: verb.communication
Definition: bring forward and present to the mind
Lemmas: ['present', 'represent', 'lay_out']

- Synset name: stage.v.01 -
POS: verb.creation
Definition: perform (a play), especially on a stage
Lemmas: ['stage', 'present', 'represent']

- Synset name: present.v.04 -
POS: verb.possession
Defin

`Synset('present.n.01')`과 `Synset('present.n.02')`는 같은 명사지만 각각 다른 의미를 지닙니다. 이처럼 synset은 단어 하나가 지닌 여러 시맨틱 정보를 개별 클래스로 나타냅니다.

`WordNet`은 단어 간 관계를 유사도로 나타낼 수 있습니다. `synsets` 객체는 단어 간 유사도를 나타내는 `path_similarity()` 메서드를 가집니다. 이 메서드를 사용하여 여러 단어의 상호 유사도를 확인해 보겠습니다.

In [ ]:
import pandas as pd

tree = wn.synset('tree.n.01')
lion = wn.synset('lion.n.01')
tiger = wn.synset('tiger.n.01')
cat = wn.synset('cat.n.01')
dog = wn.synset('dog.n.01')

entities = [tree, lion, tiger, cat, dog]
similarities = []
entity_names = [entity.name().split('.')[0] for entity in entities]

for entity in entities:
    similarity = [round(entity.path_similarity(compared_entity), 2) \
                  for compared_entity in entities]
    similarities.append(similarity)

similarity_df = pd.DataFrame(similarities, columns=entity_names, index=entity_names)
similarity_df

tree  lion  tiger   cat   dog
tree   1.00  0.07   0.14  0.08  0.12
lion   0.07  1.00   0.08  0.25  0.17
tiger  0.14  0.08   1.00  0.09  0.17
cat    0.08  0.25   0.09  1.00  0.20
dog    0.12  0.17   0.17  0.20  1.00

1에 가까울수록 상호 유사도가 높습니다.

`SentiWordNet`은 `Wordnet`의 `synsets()` 메서드와 비슷한 `senti_synsets()` 메서드를 가집니다. 사용해 보겠습니다.

In [ ]:
from nltk.corpus import sentiwordnet as swn

# 리스트 형태로 변환하지 않으면 `filter` 형태입니다.
senti_synsets = list(swn.senti_synsets('slow'))
print('senti_synsets type:', type(senti_synsets))
print('senti_synsets 값 개수:', len(senti_synsets))
print('senti_synsets 값 :', senti_synsets)

senti_synsets type: <class 'list'>
senti_synsets 값 개수: 11
senti_synsets 값 : [SentiSynset('decelerate.v.01'), SentiSynset('slow.v.02'), SentiSynset('slow.v.03'), SentiSynset('slow.a.01'), SentiSynset('slow.a.02'), SentiSynset('dense.s.04'), SentiSynset('slow.a.04'), SentiSynset('boring.s.01'), SentiSynset('dull.s.08'), SentiSynset('slowly.r.01'), SentiSynset('behind.r.03')]


`SentiSynset` 클래스를 리스트 형태로 반환했습니다. 이 클래스는 감성 지수와 객관성 지수를 출력하는 메서드 3개를 가집니다. 긍정 감성 지수를 출력하는 `pos_score()`, 부정 감성 지수를 출력하는 `neg_score()`, 객관성 지수를 출력하는 `odj_score()` 메서드입니다. 몇몇 단어에 적용해 보겠습니다.

In [ ]:
term = 'mother'
synsets = wn.synsets(term)
print('synsets 값 개수:', len(synsets))
print('synsets 값 :', synsets)
print()

for synset in synsets:
    print('- Synset name:', synset.name(), '-')
    print('POS:', synset.lexname())
    print('Definition:', synset.definition())
    print('Lemmas:', synset.lemma_names())
    print()

senti_synsets = list(swn.senti_synsets('mother'))
print('senti_synsets 값 개수:', len(senti_synsets))
print('senti_synsets 값 :', senti_synsets)
print()

mother = swn.senti_synset('mother.n.01')
print('mother.n.01 긍정 감성 지수:', mother.pos_score())
print('mother.n.01 부정 감성 지수:', mother.neg_score())
print('mother.n.01 객관성 지수:', mother.obj_score())

synsets 값 개수: 7
synsets 값 : [Synset('mother.n.01'), Synset('mother.n.02'), Synset('mother.n.03'), Synset('mother.n.04'), Synset('mother.n.05'), Synset('mother.v.01'), Synset('beget.v.01')]

- Synset name: mother.n.01 -
POS: noun.person
Definition: a woman who has given birth to a child (also used as a term of address to your mother)
Lemmas: ['mother', 'female_parent']

- Synset name: mother.n.02 -
POS: noun.substance
Definition: a stringy slimy substance consisting of yeast cells and bacteria; forms during fermentation and is added to cider or wine to produce vinegar
Lemmas: ['mother']

- Synset name: mother.n.03 -
POS: noun.person
Definition: a term of address for an elderly woman
Lemmas: ['mother']

- Synset name: mother.n.04 -
POS: noun.person
Definition: a term of address for a mother superior
Lemmas: ['mother']

- Synset name: mother.n.05 -
POS: noun.cognition
Definition: a condition that is the inspiration for an activity or situation
Lemmas: ['mother']

- Synset name: mother.v.0

In [ ]:
term = 'outstanding'
synsets = wn.synsets(term)
print('synsets 값 개수:', len(synsets))
print('synsets 값 :', synsets)
print()

for synset in synsets:
    print('- Synset name:', synset.name(), '-')
    print('POS:', synset.lexname())
    print('Definition:', synset.definition())
    print('Lemmas:', synset.lemma_names())
    print()

senti_synsets = list(swn.senti_synsets('outstanding'))
print('senti_synsets 값 개수:', len(senti_synsets))
print('senti_synsets 값 :', senti_synsets)
print()

# `'great.s.02'`와 같은 지수입니다.
outstanding = swn.senti_synset('outstanding.s.04')
print('outstanding.s.04 긍정 감성 지수:', outstanding.pos_score())
print('outstanding.s.04 부정 감성 지수:', outstanding.neg_score())
print('outstanding.s.04 객관성 지수:', outstanding.obj_score())

synsets 값 개수: 4
synsets 값 : [Synset('outstanding.s.01'), Synset('outstanding.s.02'), Synset('outstanding.s.03'), Synset('great.s.02')]

- Synset name: outstanding.s.01 -
POS: adj.all
Definition: distinguished from others in excellence
Lemmas: ['outstanding']

- Synset name: outstanding.s.02 -
POS: adj.all
Definition: having a quality that thrusts itself into attention
Lemmas: ['outstanding', 'prominent', 'salient', 'spectacular', 'striking']

- Synset name: outstanding.s.03 -
POS: adj.all
Definition: owed as a debt
Lemmas: ['outstanding', 'owing', 'undischarged']

- Synset name: great.s.02 -
POS: adj.all
Definition: of major significance or importance
Lemmas: ['great', 'outstanding']

senti_synsets 값 개수: 4
senti_synsets 값 : [SentiSynset('outstanding.s.01'), SentiSynset('outstanding.s.02'), SentiSynset('outstanding.s.03'), SentiSynset('great.s.02')]

outstanding.s.04 긍정 감성 지수: 0.75
outstanding.s.04 부정 감성 지수: 0.0
outstanding.s.04 객관성 지수: 0.25


### **5.4.2. SentiWordNet을 이용한 영화 감상평 감성 분석**

IMDB 영화 감상평 분석을 수행해 보겠습니다. `SentiWordNet`을 이용한 감성 분석은 아래 순서로 수행합니다.

1. 문서를 문장 단위로 분해합니다.

1. 다시 문장을 단어 단위로 토큰화하고 품사 태깅을 수행합니다(`WordNet` 사용).

1. 품사 태깅을 수행한 단어를 기반으로 `synset` 객체와 `senti_synset` 객체를 생성합니다.

1. `senti_synset`으로 감성 지수를 구하고 합산하여 특정 임계치 이상이면 긍정, 그렇지 않으면 부정으로 판단합니다.

`SentiWordNet`을 이용하려면 `WordNet`을 이용하여 문서를 다시 단어로 토큰화하고 어근 추출(Lemmatization)과 품사 태깅(POS Tagging)을 적용해야 합니다. 먼저 품사 태깅을 수행하는 내부 함수를 선언하겠습니다.

In [7]:
from nltk.corpus import wordnet as wn

# NLTK PennTreebank Tag를 기반으로 `WordNet` 기반 품사 Tag로 변환합니다.
def penn_to_wn(tag):
    # 형용사로 변환합니다.
    if tag.startswith('J'):
        return wn.ADJ
    # 명사로 변환합니다.
    elif tag.startswith('N'):
        return wn.NOUN
    # 부사로 변환합니다.
    elif tag.startswith('R'):
        return wn.ADV
    # 동사로 변환합니다.
    elif tag.startswith('V'):
        return wn.VERB

이제 문서를 문장에서 단어 토큰화를 수행하고 품사 태깅을 수행한 후, `SentiSynset` 클래스를 생성하고, Polarity Score를 합산하는 함수를 선언하겠습니다. 총 감성 지수가 0 이상이면 긍정, 그렇지 않으면 부정으로 예측합니다. 함수 로직은 주석으로 설명하겠습니다.

In [13]:
from nltk.stem import WordNetLemmatizer
from nltk.corpus import sentiwordnet as swn
from nltk import sent_tokenize, word_tokenize, pos_tag

def swn_polarity(text):
    # 감성 지수를 초기화합니다.
    sentiment = 0.0
    # 토큰 개수를 초기화합니다.
    tokens_count = 0

    # 단어의 원형을 찾는 인스턴스를 생성합니다.
    lemmatizer = WordNetLemmatizer()
    # 문장을 토큰화하는 인스턴스를 생성합니다.
    raw_sentences = sent_tokenize(text)
    
    # 분해된 문장별로 단어를 토큰화하고 품사 태깅을 수행한 후 `SentiSynset`을 생성하여
    # 감성 지수를 합산합니다.
    for raw_sentence in raw_sentences:
        # NLTK 기반의 품사 태깅 문장을 추출합니다.
        tagged_sentence = pos_tag(word_tokenize(raw_sentence))
        for word, tag in tagged_sentence:
            # WordNet 기반의 품사 태깅과 어근 추출을 수행합니다.
            wn_tag = penn_to_wn(tag)
            if wn_tag not in (wn.NOUN, wn.ADJ, wn.ADV):
                continue
            lemma = lemmatizer.lemmatize(word, pos=wn_tag)
            if not lemma:
                continue
            # 어근을 추출한 단어와 WordNet 기반의 품사 태깅을 입력하여 Synset 객체를 생성합니다.
            synsets = wn.synsets(lemma, pos=wn_tag)
            if not synsets:
                continue
            # `sentiwordnet`의 감성 단어 분석으로 감성 synset을 추출합니다.
            # 긍정 감성 지수는 +, 부정 감성 지수는 -로 합산하여 감성 지수를 계산합니다.
            synset = synsets[0]
            swn_synset = swn.senti_synset(synset.name())
            sentiment += (swn_synset.pos_score() - swn_synset.neg_score())
            tokens_count += 1

        if not tokens_count:
            return 0

        # 총 score가 0 이상이면 긍정 1, 그렇지 않으면 부정 0을 반환합니다.
        if sentiment >= 0:
            return 1
        return 0

이 함수들을 토대로 개별 문서에 적용하여 긍정과 부정을 예측하겠습니다. 'preds' 칼럼을 추가하여 감성 평가를 넣고 `SentiWordNet`의 감성 분석 예측 성능을 확인해 보겠습니다.

In [14]:
review_df['preds'] = review_df['review'].apply(lambda x: swn_polarity(x))
y_target = review_df['sentiment'].values
preds = review_df['preds'].values

In [ ]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score
from sklearn.metrics import recall_score, f1_score, roc_auc_score
import numpy as np

print(confusion_matrix(y_target, preds))
print('정확도:', np.round(accuracy_score(y_target, preds), 4))
print('정밀도:', np.round(precision_score(y_target, preds), 4))
print('재현율:', np.round(recall_score(y_target, preds), 4))

[[7668 4832]
 [3636 8864]]
정확도: 0.6613
정밀도: 0.6472
재현율: 0.7091


만족스러운 수치는 아닙니다.

## **5.5. VADER를 이용한 감성 분석**

**VADER**는 소셜 미디어의 감성 분석 용도로 개발된 룰 기반의 Lexicon입니다. `SentimentIntensityAnalyzer` 클래스를 이용하여 쉽게 감성 분석을 제공합니다. VADER는 NLTK 패키지의 서브 모듈로 제공되거나 단독 패키지로 제공됩니다. VADER를 NLTK의 서브 모듈로 설치하려면 위에서 수행한 `nltk.download('all')` 코드를 취하면 됩니다. 별도 모듈로 사용하려면 `pip install vaderSentiment` 코드를 취하고, `from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer` 코드로 클래스를 임포트합니다.

이제 VADER의 사용법을 살펴보겠습니다. `nltk` 패키지에 포함된 서브 모듈 `vader`의 `SentimentIntensityAnalyzer` 클래스를 사용하겠습니다.

In [20]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

senti_analyzer = SentimentIntensityAnalyzer()
# `polarity_scores()` 메서드로 감성 점수를 반환합니다.
senti_scores = senti_analyzer.polarity_scores(review_df['review'][0])
print(senti_scores)

{'neg': 0.13, 'neu': 0.743, 'pos': 0.127, 'compound': -0.7943}


'neg'는 부정 감성 지수, 'neu'는 중립 감성 지수, 'pos'는 긍정 감성 지수입니다. 'compound'는 세 지수를 조합하여 -1 ~ 1 사이의 감성 지수를 표현한 값입니다. 0.1 이상이면 긍정이고, 그렇지 않으면 부정으로 판단하지만 상황에 따라 조정하여 예측 성능을 조절할 수 있습니다.

VADER를 사용하여 IMDB 감성 분석을 수행해 보겠습니다. `vader_polarity()` 함수를 선언하여 감성 분석에 활용하겠습니다. 함수 로직에 대한 설명은 주석을 토대로 하겠습니다.

In [21]:
# 영화 감상평 텍스트(`review`), 긍정과 부정을 결정하는 임곗값(`threshold`)을 받습니다.
def vader_polarity(review, threshold=0.1):
    analyzer = SentimentIntensityAnalyzer()
    scores = analyzer.polarity_scores(review)

    # 'compound' 값을 토대로 `threshold` 파라미터보다 크면 1, 그렇지 않으면 0을 반환합니다.
    agg_score = scores['compound']
    final_sentiment = 1 if agg_score >= threshold else 0
    return final_sentiment

이 함수들을 토대로 개별 문서에 적용하여 긍정과 부정을 예측하겠습니다. 'vader_preds' 칼럼을 추가하여 감성 평가를 넣고, 이 감성 평가로 VADER의 예측 성능을 확인하겠습니다.


In [23]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score
from sklearn.metrics import recall_score, f1_score, roc_auc_score
import numpy as np

review_df['vader_preds'] = review_df['review'].apply(lambda x: vader_polarity(x, 0.1))
y_target = review_df['sentiment'].values
vader_preds = review_df['vader_preds'].values

print(confusion_matrix(y_target, vader_preds))
print('정확도:', np.round(accuracy_score(y_target, vader_preds), 4))
print('정밀도:', np.round(precision_score(y_target, vader_preds), 4))
print('재현율:', np.round(recall_score(y_target, vader_preds), 4))

[[ 6736  5764]
 [ 1867 10633]]
정확도: 0.6948
정밀도: 0.6485
재현율: 0.8506


`SentiWordNet`보다 전반적으로 향상됐고 특히 재현율은 높은 수치로 올랐습니다. 지도학습을 기반으로 한 감성 분석보다는 낮은 성능을 보이지만 결정 클래스 값이 주어지지 않은 상황을 고려하면 적당히 만족할 만한 성능이라고 판단할 수 있습니다.

# **6. 토픽 모델링(Topic Modeling) - 20 뉴스그룹**

**토픽 모델링**은 문서 집합에 숨어 있는 주제를 찾아내는 작업입니다. 사람이 수행하는 토픽 모델링은 더 함축적인 의미로 문장을 요약하는 것뿐이지만 ML 기반 토픽 모델링은 숨어 있는 주제를 효과적으로 표현할 수 있는 중심 단어를 함축적으로 추출합니다.

ML 기반 토픽 모델링에 자주 사용되는 기법은 **LSA(Latent Semantic Analysis)**와 **LDA(Latent Dirichlet Allocation)**입니다. 위에서 사용했던 20 뉴스그룹 데이터를 사용하여 LSA를 토대로 토픽 모델링을 수행해 보겠습니다. 20 뉴스그룹 데이터가 가진 주제 20가지 목록(`target_names` 속성)은 아래와 같습니다.

```
['alt.atheism', 'comp.graphics', 'comp.os.ms-windows.misc', 'comp.sys.ibm.pc.hardware', 'comp.sys.mac.hardware', 'comp.windows.x', 'misc.forsale', 'rec.autos', 'rec.motorcycles', 'rec.sport.baseball', 'rec.sport.hockey', 'sci.crypt', 'sci.electronics', 'sci.med', 'sci.space', 'soc.religion.christian', 'talk.politics.guns', 'talk.politics.mideast', 'talk.politics.misc', 'talk.religion.misc']
```

여기에서 8가지 주제를 추출하고 텍스트에 LDA 기반 토픽 모델링을 적용하겠습니다. LDA는 카운트 기반 벡터화만 사용하므로 카운트 기반으로 벡터화하겠습니다.

In [24]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# 모토사이클, 야구, 그래픽스, 윈도우, 중동, 기독교, 전자공학, 의학 주제를 추출합니다.
cats = ['rec.motorcycles', 'rec.sport.baseball', 'comp.graphics', 'comp.windows.x',
        'talk.politics.mideast', 'soc.religion.christian', 'sci.electronics', 'sci.med']
news_df = fetch_20newsgroups(subset='all', remove=('headers', 'footers', 'quotes'),
                             categories=cats, random_state=0)

count_vect = CountVectorizer(max_df=0.95, max_features=1000, min_df=2, stop_words='english',
                             ngram_range=(1, 2))
feat_vect = count_vect.fit_transform(news_df.data)
print('CountVectorizer Shape:', feat_vect.shape)

CountVectorizer Shape: (7862, 1000)


문서 7862개에 피처 1000개로 구성된 행렬 데이터입니다. 이 피처 벡터화가 수행된 데이터 세트를 기반으로 LDA 토픽 모델링을 수행합니다. 토픽 개수는 8개로 정할 것이므로 `LatentDirichletAllocation` 클래스의 `n_components` 파라미터에 `8`을 지정하겠습니다.


In [25]:
lda = LatentDirichletAllocation(n_components=8, random_state=0)
lda.fit(feat_vect)

LatentDirichletAllocation(n_components=8, random_state=0)

위 코드 과정을 통해 선언한 인스턴스는 `components_` 속성을 갖습니다. 이 속성은 토픽별로 각 단어 피처가 얼마나 많이 그 토픽에 할당됐는지를 의미하는 수치를 가집니다. 높은 값일수록 해당 단어 피처는 그 토픽의 중심 단어가 됩니다. 속성을 확인해 보겠습니다.

In [29]:
print(lda.components_)
print(lda.components_.shape)

[[3.60992018e+01 1.35626798e+02 2.15751867e+01 ... 3.02911688e+01
  8.66830093e+01 6.79285199e+01]
 [1.25199920e-01 1.44401815e+01 1.25045596e-01 ... 1.81506995e+02
  1.25097844e-01 9.39593286e+01]
 [3.34762663e+02 1.25176265e-01 1.46743299e+02 ... 1.25105772e-01
  3.63689741e+01 1.25025218e-01]
 ...
 [3.60204965e+01 2.08640688e+01 4.29606813e+00 ... 1.45056650e+01
  8.33854413e+00 1.55690009e+01]
 [1.25128711e-01 1.25247756e-01 1.25005143e-01 ... 9.17278769e+01
  1.25177668e-01 3.74575887e+01]
 [5.49258690e+01 4.47009532e+00 9.88524814e+00 ... 4.87048440e+01
  1.25034678e-01 1.25074632e-01]]
(8, 1000)


다만 이 속성 값만으로 토픽별 단어 연관도를 확인하기 어렵습니다. 새 함수를 선언하여 토픽별로 연관도가 높은 단어를 순서대로 나타내 보겠습니다.

---

정리 중

---

# **7. 문서 군집화 소개와 실습(Opinion Review 데이터 세트)**

## **7.1. 문서 군집화 개념**

## **7.2. Opinion Review 데이터 세트를 이용한 문서 군집화 수행하기**

## **7.3. 군집별 핵심 단어 추출하기**

# **8. 문서 유사도**

## **8.1. 문서 유사도 측정 방법 - 코사인 유사도**

## **8.2. 두 벡터 사잇각**

## **8.3. Opinion Review 데이터 세트를 이용한 문서 유사도 측정**

# **9. 한글 텍스트 처리 - 네이버 영화 평점 감성 분석**

## **9.1. 한글 NLP 처리의 어려움**

## **9.2. KoNLPy 소개**

## **9.3. 데이터 로딩**

# **10. 텍스트 분석 실습 - 캐글 Mercari Price Suggestion Challenge**

## **10.1. 데이터 전처리**

## **10.2. 피처 인코딩과 피처 벡터화**

## **10.3. 릿지 회귀 모델 구축 및 평가**

## **10.4. LightGBM 회귀 모델 구축과 앙상블을 이용한 최종 예측 평가**

# **11. 정리**